core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/fuse/PlanNodeFuser.java

```diff
@@ -0,0 +1,433 @@
+            left = context.getLookup().resolve(((ProjectNode) left).getSource());
+        }
+        while (right instanceof ProjectNode && outputsSameAsSource((ProjectNode) right)) {
+            // ignore identity projection
+            right = context.getLookup().resolve(((ProjectNode) right).getSource());
+        }
+
+        if (left instanceof FilterNode || right instanceof FilterNode) {
+            return fuseFilter(left, right);
+        }
+
+        if (!left.getClass().equals(right.getClass())) {
+            return Optional.empty();
+        }
+        if (left instanceof TableScanNode) {
+            return fuseTableScan((TableScanNode) left, (TableScanNode) right);
+        }
+        if (left instanceof AggregationNode) {
+            return fuseAggregation((AggregationNode) left, (AggregationNode) right);
+        }
+
+        return Optional.empty();
+    }
+
+    private Optional<FusedPlanNode> fuseFilter(PlanNode left, PlanNode right)
+    {
+        if (left instanceof FilterNode leftFilter && right instanceof FilterNode rightFilter) {
+            return fuseFilter(leftFilter.getSource(), leftFilter.getPredicate(), rightFilter.getSource(), rightFilter.getPredicate());
+        }
+        if (left instanceof FilterNode leftFilter) {
+            return fuseFilter(leftFilter.getSource(), leftFilter.getPredicate(), right, TRUE_LITERAL);
+        }
+        if (right instanceof FilterNode rightFilter) {
+            return fuseFilter(left, TRUE_LITERAL, rightFilter.getSource(), rightFilter.getPredicate());
+        }
+
+        throw new IllegalArgumentException(String.format("expected at least one FilterNode but got %s and %s", left, right));
+    }
+
+    private Optional<FusedPlanNode> fuseFilter(PlanNode leftSource, Expression leftPredicate, PlanNode rightSource, Expression rightPredicate)
+    {
+        Optional<FusedPlanNode> maybeFusedSource = fuse(leftSource, rightSource);
+        if (maybeFusedSource.isEmpty()) {
+            return Optional.empty();
+        }
+        FusedPlanNode fusedSource = maybeFusedSource.get();
+
+        Expression mappedRightPredicate = fusedSource.symbolMapping().map(rightPredicate);
+        if (leftPredicate.equals(mappedRightPredicate)) {
+            // simplified version. since the predicates match, and are applied in the fused FilterNode as predicate,
+            // they don't have to be added to the left and right filters.
+            return Optional.of(new FusedPlanNode(
+                    new FilterNode(
+                            context.getIdAllocator().getNextId(),
+                            fusedSource.plan(),
+                            leftPredicate),
+                    fusedSource.symbolMapping(),
+                    fusedSource.leftFilter(),
+                    fusedSource.rightFilter()));
+        }
+
+        PlanNode fused;
+        if (leftPredicate.equals(TRUE_LITERAL) || mappedRightPredicate.equals(TRUE_LITERAL)) {
```

Comment:
I think instead of `||` we should have `&` here. Let me know if I'm missing something.

Reply:
Basically what if right predicate isn't TRUE while left is?

Reply:
Fusing `Filter` basically pushes the left and right conditions to the "fused" conditions but it also has to add `left or right` to the plan to filter out rows that do not match any side.
If either left or right condition is `TRUE` then `left or right` also equals `TRUE` so we don't have to add the filter.
This means basically that one side is full scan and so we need to push all rows anyway.
Hope that helps. Do you think it's with extending the comment there?

Reply:
This makes sense. Yeah I think we can add a small comment here

Reply:
comment extended